crates/ra_hir/src/ty.rs

```diff
@@ -660,66 +694,121 @@ pub(crate) fn type_for_enum_variant(db: &impl HirDatabase, ev: EnumVariant) -> T
     type_for_enum(db, enum_parent)
 }
 
-pub(super) fn type_for_def(db: &impl HirDatabase, def_id: DefId) -> Ty {
-    let def = def_id.resolve(db);
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum TypableDef {
```

Comment:
A comment explaining what `Typable` means would be good. I'm assuming it's something that has a type? Or it is something a user can type?